src/containerbuddy/etcd.go

```diff
@@ -0,0 +1,240 @@
+package main
+
+import (
+	"fmt"
+	"log"
+	"sort"
+	"time"
+
+	"encoding/json"
+
+	"github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/context"
+	"github.com/coreos/etcd/client"
+)
+
+// Etcd is a service discovery backend for CoreOS etcd
+type Etcd struct {
+	Client client.Client
+	API    client.KeysAPI
+	Prefix string
+}
+
+// EtcdServiceNode is an instance of a service
+type EtcdServiceNode struct {
+	ID      string   `json:"id"`
+	Name    string   `json:"name"`
+	Address string   `json:"address"`
+	Port    int      `json:"port"`
+	Tags    []string `json:"tags,omitempty"`
+}
+
+// NewEtcdConfig creates a new service discovery backend for etcd
+func NewEtcdConfig(config map[string]interface{}) Etcd {
+	etcd := Etcd{
+		Prefix: "/containerbuddy",
+	}
+	etcdConfig := client.Config{}
+	switch endpoints := config["endpoints"].(type) {
+	case string:
+		etcdConfig.Endpoints = []string{endpoints}
+	case []string:
+		etcdConfig.Endpoints = endpoints
+	default:
+		log.Fatal("Must provide etcd endpoints")
+	}
+
+	prefix, ok := config["prefix"].(string)
+	if ok {
+		etcd.Prefix = prefix
+	}
+
+	etcdClient, err := client.New(etcdConfig)
+	if err != nil {
+		log.Fatal(err)
+	}
+	etcd.Client = etcdClient
+	etcd.API = client.NewKeysAPI(etcdClient)
+	return etcd
+}
+
+// Deregister removes this instance from the registry
+func (c Etcd) Deregister(service *ServiceConfig) {
+	c.deregisterService(service)
+}
+
+// MarkForMaintenance removes this instance from the registry
+func (c Etcd) MarkForMaintenance(service *ServiceConfig) {
+	c.deregisterService(service)
+}
+
+// SendHeartbeat refreshes the TTL of this associated etcd node
+func (c Etcd) SendHeartbeat(service *ServiceConfig) {
+	if c.checkServiceExists(service) {
```

Comment:
According to https://godoc.org/github.com/coreos/etcd/client#KeysAPI the `Set` endpoint returns an error. Rather than making a call to test for existence and then making a second call, we should just make the `Set` call and catch the error, falling back to `registerService` if we get the error.